# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Arryan-56/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Semantic Clustering & Topic Mapping. I'm choosing this lane because grouping pages by topic similarity (rather than exact keyword match) surfaces real content problems — overlapping pages competing for the same topic, and gap topics with search demand but no dedicated page. This maps directly to a content strategy decision a team can act on, and it lets me apply embeddings and clustering to a genuinely business-relevant problem over the next 7 weeks, rather than a purely academic exercise.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


Search question: Which pages cover the same topic even when worded differently, and which topics have demand but no page? Unit of analysis: a single page on a client's site. Output: a topic cluster per page + list of overlap topics and gap topics. Decision it improves: whether to merge competing pages or create new content for underserved topics. Action someone takes: SEO/content team merges overlapping pages, or greenlights new content for a gap topic. Cost of a wrong call: false overlap → merging/deleting content that served a real distinct search intent, losing traffic; false gap → wasted effort building content nobody searches for. Why ML helps: manually judging topic similarity across hundreds of pages doesn't scale; embeddings + clustering surface patterns for a human to validate — not full automation.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [1]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/Arryan-56/flyrank-ml-internship/main/data/raw/content_refresh_anonymized.csv')

print("Total pages:", len(df))
print("Unique clients:", df.client_id.nunique())
print("Avg pages per client:", round(len(df)/df.client_id.nunique(), 1))
print("Clients with 100+ pages:", (df.client_id.value_counts() > 100).sum())
print(df.content_type.value_counts())
print(df.main_intent.value_counts())

Total pages: 30000
Unique clients: 32
Avg pages per client: 937.5
Clients with 100+ pages: 24
content_type
keyword article       27207
feedly article         2096
comparison article      697
Name: count, dtype: int64
main_intent
informational    17235
transactional     5733
commercial        4612
navigational        46
Name: count, dtype: int64


Real numbers observed: 30,000 total pages across 32 clients — averaging ~938 pages per client. 24 of 32 clients have 100+ pages each (one client has 7,008 pages). 27,207 pages (91%) are the same content_type ("keyword article"), and intent is dominated by "informational" (17,235) — meaning many pages plausibly compete for similar topics.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


What I can claim: based on this anonymized sample, there is enough page-per-client volume (avg ~938 pages, some clients with 7,000+) that manual topic review isn't practical, and the dominance of one content_type/intent suggests real potential for topic overlap. Any cluster output will need human review before action — this is decision-support, not an automated ranking or algorithm-prediction tool. What I can't claim: I cannot claim to predict Google's ranking algorithm, guarantee traffic gains from merging pages, or say clustering results reflect actual search intent without human validation.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.